### PR TITLE
Add fields 'ttl' and 'num_uses' to SecretID generation.

### LINKED ISSUE #14390: Customizing num_uses and ttl per approle secret id
**Is your feature request related to a problem? Please describe.**
When I generate a new secret ID for an approle, I would like to customize the TTL and/or num_uses for that secret ID, similarly to how I can customize the [cidr_list](https://www.vaultproject.io/api-docs/auth/approle#cidr_list) and [token_bound_cidrs](https://www.vaultproject.io/api-docs/auth/approle#token_bound_cidrs-1). 

**Describe the solution you'd like**
Additional API parameters for the [approle secret ID creation](https://www.vaultproject.io/api-docs/auth/approle#generate-new-secret-id) allowing custom values to be set for `num_uses` and `ttl`.

**Describe alternatives you've considered**
Since `num_uses` and `ttl` are currently tied to the role configuration, any use case for an approle that requires a different value for either parameter than the default means I have to create an entirely new approle with identical policies but different configuration. This is possible, but tedious.

**Explain any additional use-cases**
In particular, this can be useful for the recommended [approle use pattern that involves wrapping](https://learn.hashicorp.com/tutorials/vault/pattern-approle?in=vault/recommended-patterns). For example, it would be very convenient to specify the `num_uses` for a new secret ID being passed to automation, even if the approle configuration did not have `num_uses` set.

### PR DESCRIPTION
**Pull request by RemcoBuddelmeijer**
Add the fields 'ttl' and 'num_uses' to the endpoints `/auth/approle/role/<role_name>/secret-id` and `/auth/approle/role/<role_name>/custom-secret-id`.
Edited the help synopsis and description to refer to these fields. Both for the endpoints respectively but also for the `secret-id-num-uses` and `secret-id-ttl` endpoints.

Fixes #14390

### CODE REVIEW

**Comment by swayne275** [comment]
thanks! it appears we have some [formatting issues](https://app.circleci.com/pipelines/github/hashicorp/vault/29149/workflows/a6378d19-dd61-4796-b9ee-00e620c431db/jobs/404062)

**Comment by swayne275** [comment]
sure, give it a shot and we can take a deeper look

**Comment by RemcoBuddelmeijer** [comment]
The test seem to be passing now 😄

**Comment by kent007** [comment]
Hi everyone,

Just want to keep the conversation moving on this feature. I see Vault 1.10 just rolled out, any idea when these changes might be generally released?

File: builtin/credential/approle/path_role.go
Comment (hghaf099) [inline builtin/credential/approle/path_role.go:2381]: Please move this check to line 2388, as this is not doing much here.

File: builtin/credential/approle/path_role.go
Comment (hghaf099) [inline builtin/credential/approle/path_role.go:2418]: Is the int(...) casting necessary?

Comment (hghaf099) [inline builtin/credential/approle/path_role.go:2418]: Is the int(...) casting necessary?

File: builtin/credential/approle/path_role_test.go
Comment (hghaf099) [inline builtin/credential/approle/path_role_test.go:1202]:
```suggestion
		t.Fatalf("secret_id_num_uses has not been set by the 'num_uses' field")
```

Comment (hghaf099) [inline builtin/credential/approle/path_role_test.go:1202]:
```suggestion
		t.Fatalf("secret_id_ttl has not been set by the 'num_uses' field")
```
```suggestion
		t.Fatalf("secret_id_num_uses has not been set by the 'num_uses' field")
```

File: builtin/credential/approle/path_role_test.go
Comment (hghaf099) [inline builtin/credential/approle/path_role_test.go:1222]:
```suggestion
		t.Fatalf("secret_id_num_uses has not been set by the 'num_uses' field")
```

Comment (hghaf099) [inline builtin/credential/approle/path_role_test.go:1222]:
```suggestion
		t.Fatalf("secret_id_num_uses has not been set by the 'num_uses' field")
```

File: changelog/14474.txt
Comment (hghaf099) [inline changelog/14474.txt:2]: would you please be more elaborative here? Like to indicate that the two fields are per secretID, yet if they are not configured, the role specific values are used.

Comment (hghaf099) [inline changelog/14474.txt:2]: would you please be more elaborative here? Like to indicate that the two fields are per secretID, yet if they are not configured, the role specific values are used.

File: website/content/api-docs/auth/approle.mdx
Comment (hghaf099) [inline website/content/api-docs/auth/approle.mdx:527]: can we keep the changes to what has actually changed please?

Comment (hghaf099) [inline website/content/api-docs/auth/approle.mdx:527]: can we keep the changes to what has actually changed please?

File: builtin/credential/approle/path_role.go
Comment (hghaf099) [inline builtin/credential/approle/path_role.go:2386]: After some internal discussions, we have decided that we would like to bound the per SecretID num_uses and ttl by the corresponding values in the Role parameters. Like, if we set the num_uses for a secretID to 200 and the role. SecretIDNumUses is 100, the value for the role should be used. And the same for the ttl.

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: You are correct, this is not necessary. Removed :)

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2381]: I have moved it to line 2382 instead. This check is only to validate the supplied field, not the role's configuration. Moved it instead of creating a second duplicate variable on the stack from `numUsesRaw`.

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: Wouldn't this defeat the purpose of the feature request that was filed? Here it is specifically requested to separate the two and in particular, be able to override the role's configuration.
Without the possibility to override the role's configuration you would still have to create a separate role.

File: changelog/14474.txt
Comment (RemcoBuddelmeijer) [reply changelog/14474.txt:2]: I'll change this once the rest of the conversations are resolved :)

File: website/content/api-docs/auth/approle.mdx
Comment (RemcoBuddelmeijer) [reply website/content/api-docs/auth/approle.mdx]: I didn't think too much of this, but changed and pushed!

File: builtin/credential/approle/path_role.go
Comment (kent007) [reply builtin/credential/approle/path_role.go]: This seems consistent with the existing rules for CIDR block restrictions...if the role value is set, then the per secret-id values must be a subset. Likewise, if the role value is _not_ set, then the per secret-id values are unbounded. I think this would still meet the original feature request, since the user can choose to set maximum values as part of the role config (or leave them unset and allow any value per-secret ID)

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: > The override is still possible though, the only constraint is that the new secretID ttl/num_uses could not be more than the corresponding ones in the role config. So, this way, the Role should be configured with highest required values for those parameters, and this way, the secretIDs could be configured to have different values but in a defined range.

Okay this makes more sense to me. While I still would disagree with a couple of points, it looks like the original requestee has given his stamp of approval.
I'll think of a solution that would fit best for the situation.

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: @hghaf099 @kent007 After a short consiledation, I came to the conclusion to simply lower bind the SecretID's ttl and num_uses to the Role's ttl and num_uses. This yields to a PR without any changes outside of the scope on the Role's configuration, possibly requiring users to change the approach they take when creating a Role.
For the best result possible I'll go for a change where validation is performed rather than implicitly overriding the SecretId's defined ttl/num_uses with the Role's ttl/num_uses.

Comment (kent007) [reply builtin/credential/approle/path_role.go]: Hi @RemcoBuddelmeijer, I'm wondering what else remains to be done for this implementation?  It sounds like we just need a few additional checks on the values presented at secret-id creation time...I have some time available this month to finish it  up if need be

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: Hey @kent007 I was unavailable the past days but I have the weekend planned to finish up the implementation!

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: Intermediate update. My days have been packed far more than expected. Right now I am out of the country and expecting to be back next weekend.

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go]: Updated the branch with the new requirements. Updated the upstream and merged main back into this branch. Will change changelog. Please feel free to review!

File: builtin/credential/approle/path_role.go
Comment (kent007) [inline builtin/credential/approle/path_role.go:2400]: Wouldn't this scenario make sense though? If the default TTL on the role was 90 days, and the client requested a secret-id with a TTL of 5 minutes for some process, then the shorter TTL should win. Just like num_uses, the role config would serve as the upper bound?

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2400]: You are 100% correct. I’ve gone ahead and change it.

**Comment by RemcoBuddelmeijer** [comment]
Hello all! I was hoping to bump up this PR to get this looked at and possibly  merged. Thank you! 😄

**Comment by kent007** [comment]
Hi everyone, also bumping this PR. Would love to see this feature merged.

File: builtin/credential/approle/path_role_test.go
Comment (raskchanky) [inline builtin/credential/approle/path_role_test.go:1185]:
```suggestion
			name:    "finite num_uses ttl",
```

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2388]: > This error message confuses me a little, specifically for the first case. If numUses == 0 && role.SecretIDNumUses > 0 then num_uses isn't higher than the role's secret_id_num_uses, is it? It's lower.

Whenever either num_uses or ttl is set to zero it equals to [unlimited/infinite](https://www.vaultproject.io/api-docs/auth/approle#:~:text=A%20value%20of%20zero%20will%20allow%20unlimited%20uses.). This guard statement's first part ensures that no unlimited custom num_uses can be used whenever the role's secret id configuration uses a limited value. The second part ensures the normal flow, as long as the role's secret id configuration does not use an unlimited num_uses value.
The same applies to the ttl parameter.

> Additionally, wouldn't this get triggered by default, e.g. if num_uses isn't set (because it would default to 0)?

No this wouldn't get triggered by default. The `data#GetOk` method checks if a parameter was specified, if not, it defaults to the Role's Secret ID num_uses. This was done to not break the current implementation. The same applies to the ttl parameter.

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2397]: This piece of code was duplicated from the current approach taken to calculate the Role's Secret ID's ttl: https://github.com/hashicorp/vault/blob/22346ba24980073f8243aa5a1ee35dc18427da49/builtin/credential/approle/path_role.go#L966

Seeing that this is no longer valid, do we want to change this behaviour as well? The code was duplicated only with the intention to not change the way TTL is calculated for the SecretID.

File: builtin/credential/approle/path_role_test.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role_test.go:1144]: This was removed as the operation never changes for this request. It's not directly related to the PR, but more so a cleanup so it can be rollbacked if it's out of scope.

File: builtin/credential/approle/path_role.go
Comment (raskchanky) [reply builtin/credential/approle/path_role.go:2388]: Ah, I'm following you now, thanks. I do think the comment might be incorrect though, as it states that an error should be thrown if the role's secret_id_num_uses is higher than the specified num_uses, but the code is opposite of that (and I think in this case the code is correct).

File: builtin/credential/approle/path_role.go
Comment (raskchanky) [reply builtin/credential/approle/path_role.go:2397]: We don't want to change the behavior of the existing API, at least not in this PR. It's odd that the implementation you point out doesn't match the docs (which says it takes either a duration string or an integer). Regardless, since this is a new API, we should use `parseutil.ParseDurationSecond()`, since that will handle both duration strings and integers transparently.

File: builtin/credential/approle/path_role_test.go
Comment (raskchanky) [reply builtin/credential/approle/path_role_test.go:1144]: Ah, I see what you mean now

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2397]: I would argue that this isn't a new API. We are currently modifying an already existing API endpoint with additional new fields. It would be odd to change this now as it defaults to the Role's Secret ID Configuration, which is calculated using the invalid approach.
How about we immediatelly create a new issue and link it after this PR?

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2388]: You are correct. I will port this together while rephrasing. This was already part of the plan as it wasn't sure yet before what the approach would be. 😄

File: builtin/credential/approle/path_role.go
Comment (raskchanky) [reply builtin/credential/approle/path_role.go:2397]: You're right, I misspoke. This is not a new API. It's a new field that we're adding to an existing API. I'm arguing that since it's a new field, the value we extract from it should be parsed using the more flexible approach, that allows for either integers or duration strings. The fact that the existing TTL field doesn't do that is a bug, as far as I can tell, especially since the documentation indicates that it should. It doesn't make sense to me to merge this PR with a bug in it, when we can correct it before it's merged.

Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2397]: While I agree, now we could have a potential mismatch in the way that TTL is parsed and returned, just to point that out. Since it's obviously not a good thing to leave bugs in, I will create an issue for the existing TTL fields. In the mean time I will patch the behaviour of the new field and we can go from there!

Comment (raskchanky) [reply builtin/credential/approle/path_role.go:2397]: > This is a major change. We should be careful about this.

I disagree that it's a major change, since I'm not advocating for changing the existing behavior of anything. Everything that existed before this PR will continue to work the same way it always has. All we're doing is determining how to parse a brand new field that no one has used before.

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2577]: I will remove the ", but" clause and only specify that this option may be overriden by the ttl field

File: builtin/credential/approle/path_role.go
Comment (RemcoBuddelmeijer) [reply builtin/credential/approle/path_role.go:2569]: I will remove the ", but" clause and only specify that this option may be overriden by the num_uses field. Also did some more rephrasing.

**Comment by RemcoBuddelmeijer** [comment]
Hello everyone, periodic bump up of the PR to see if it can get merged. Thank you!

**Review by HridoyRoy — APPROVED** [review]
Lgtm

**Pull request merged**
